DDF-3172 bad.files and bad.file.extensions in system.properties is not being respected

johnsliao opened the pull request:
#### What does this PR do?
- Fix the issue where mime type improperly used the original filename and overwrote the sanitized filename.
- Fix a bug where if `files.bad`, `files.bad.extension` in `system.properties` contained any capitalization, to-be sanitized files would not be properly identified
- Minor refactoring of `generateMetacardAndContentItems()`
- Add ability to ignore files (`ignore.files` in `system.properties`)
- Update documentation as needed


#### Who is reviewing it? 
@ahoffer 
@brendan-hofmann 
@emanns95 
@AzGoalie 
@jhunzik 
@mweser 

#### Select relevant component teams: 
https://github.com/orgs/codice/teams
#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@coyotesqrl
@stustison
#### How should this be tested? (List steps with links to updated documentation)

1. Build ddf
2. Add a  dummy filename to `ignore.files` in `etc/system.properties`
3. Stand up ddf
4. Ingest the attached test files
5. `crossdomain.xml` and `thUmbs.db` will be successfully ingested as `file.bin`
6. `bad_file.cgi` will be successfully ingested as `bad_file.bin`
7. Try to ingest a file with that filename added in step 2. It will not be successful

[DDF-3172.TestFiles.zip](https://github.com/codice/ddf/files/1180825/DDF-3172.TestFiles.zip)

#### Any background context you want to provide?

#### What are the relevant tickets?
DDF-2151
[](https://codice.atlassian.net/browse/)
#### Screenshots (if appropriate)
#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests
#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/codice/ddf?pullRequest=2165) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/codice/ddf?pullRequest=2165) before we can accept your contribution.<br/><hr/>**John Liao** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/codice/ddf?pullRequest=2165) it.</sub>

ahoffer commented on platform/util/platform-util/src/test/java/org/codice/ddf/platform/util/InputValidationTest.java:
```diff
@@ -44,14 +46,21 @@
+
     @Before
     public void setup() {
         System.setProperty("bad.files",
```
I recommend using this library to manage system properties for testing. 
http://stefanbirkner.github.io/system-rules/

Using it ensures that the system properties are restored to their original values.

johnsliao replied:
Thanks for the feedback. Added

johnsliao commented:
Pushed change to use http://stefanbirkner.github.io/system-rules/ instead of `System.setProperty()` in the unit tests

johnsliao commented:
build now

mweser commented on distribution/ddf-common/src/main/resources/etc/system.properties:
```diff
@@ -127,6 +127,8 @@ bad.files=crossdomain.xml,clientaccesspolicy.xml,.htaccess,.htpasswd,hosts,passw
 bad.mime.types=text/html,text/javascript,text/x-javascript,application/x-shellscript,text/scriptlet,application/x-msdownload,application/x-msmetafile
 
+# Files uploaded matching these file names will be excluded from being stored
+ignore.files=.DS_Store
```
❓ Were there DS_store files gumming up the works to need this ignore?

emanns95 replied:
Yes right now the CDM ingests those and it's pretty annoying

mweser commented on distribution/docs/src/main/jdocs/content/_configuring/configuring-global-settings-contents.adoc:
```diff
@@ -372,7 +372,8 @@ nfs.conf,
 web.config,
-robots.txt
+robots.txt,
+Thumbs.db
```
❓ What is potentially dangerous about Thumbs.db? Just general curiosity.

bdeining replied:
It's not dangerous, but on Windows machines I believe this gets auto generated and will continuously be created and ingested through the CDM

stustison replied:
We should probably move Thumbs.db to the "ignored" list.

emanns95 replied:
Looks like the docs still need to be updated to move this to the ignored list

mweser commented on platform/util/platform-util/pom.xml:
```diff
@@ -70,6 +70,12 @@
+            <artifactId>system-rules</artifactId>
+            <version>1.16.1</version>
+            <scope>test</scope>
+        </dependency>
```
❓ Do we use this external dependency elsewhere? I know we're usually pretty leery about this sort of thing.

johnsliao replied:
I took a look. 

This dependency exists @`/platform/security/sts/security-sts-ldaplogin/pom.xml`. Are there drawbacks on moving it to a common `pom.xml`?

emanns95 replied:
You could move the version up to platform's `pom` as a property to ensure they stay on the same version (if we want them to). I'm not sure we'd want to move that dependency up but I'm not sure about the pros/cons of that.

mweser replied:
@coyotesqrl or @lessarderic might have better direction on this.

coyotesqrl replied:
As it's only used in scope-test, I see no reason not to put it in the root pom.

mweser commented on platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/InputValidation.java:
```diff
@@ -31,14 +31,22 @@
     private static final String DEFAULT_FILE = "file" + DEFAULT_EXTENSION;
 
     private static final List<String> BAD_FILES = Arrays.asList(System.getProperty("bad.files")
+            .toLowerCase()
```
❓ Was there an issue with some file names not being all lowercase? Why does this need to happen?

johnsliao replied:
Yes. If there is capitalization in the bad.files/bad.file.extensions, they wouldn't have been matched properly.

e.g. a filename as `thumbs.db` would not be matched with `Thumbs.db` in system.properties.

Although I suppose using regex would be the cleanest solution. It's just that the legacy code identified them by converting filenames to lowercase.

brendan-hofmann replied:
Regex would be nice because we could ignore all hidden files on *nix systems by default. `\..*`

mweser commented on platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/InputValidation.java:
```diff
@@ -97,4 +105,22 @@ public static boolean checkForClientSideVulnerableMimeType(String mimetype) {
+     * @param filename
+     * @return true if the filename is acceptable
+     */
+    public static boolean checkForClientSideVulnerableFileName(String filename) {
```
✏️  Might this method name benefit from following the naming convention of starting with a form of "to be", as in `isClientSideVulnerableFileName`? Might help with readability by future developers where this is used so they know it is a boolean return.

johnsliao replied:
Good point. Thanks.

mweser commented on platform/util/platform-util/src/test/java/org/codice/ddf/platform/util/InputValidationTest.java:
```diff
@@ -44,15 +48,30 @@
+                    "text/html,text/javascript,text/x-javascript,application/x-shellscript,text/scriptlet,application/x-msdownload,application/x-msmetafile");
+
+    @ClassRule
+    public static final ProvideSystemProperty ignoreFilesSystemProperties =
```
❓ Are these classrules applied under the hood for the rest of the unit tests?

johnsliao replied:
Yes, for all the `InputValidationTest.java` unit tests. If I understand your question correctly.

I thought this was helpful:
https://stackoverflow.com/questions/41121778/junit-rule-and-classrule

mweser commented on platform/util/platform-util/src/test/java/org/codice/ddf/platform/util/InputValidationTest.java:
```diff
@@ -95,4 +114,30 @@ public void testCheckForClientSideVulnerableMimeTypeGood() {
+    @Test
+    public void testSanitizeFilenameKnownBadCaseInsensitive() {
+        String sanitizedName = InputValidation.sanitizeFilename(BAD_FILE3);
+        assertThat(sanitizedName, is(DEFAULT_FILE));
```
✏️ Perhaps use the Hamcrest `assertThat` method which takes a String reason and the boolean check to help provide better logging in case of failure. Same with the other tests.

emanns95 replied:
Could also use `assertThat("This is my failure reason", sanitizedName, is(DEFAULT_FILE));` if you still want to use matchers

johnsliao replied:
Good point. Thanks

emanns95 commented on catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/OperationsMetacardSupport.java:
```diff
@@ -94,15 +94,20 @@ void generateMetacardAndContentItems(List<ContentItem> incomingContentItems,
-                    String sanitizedFilename =
-                            InputValidation.sanitizeFilename(contentItem.getFilename());
+                    if (!InputValidation.checkForClientSideVulnerableFileName(fileName)) {
+                        throw new IngestException("Ignored filename found.");
```
Maybe I'm losing some context in the scope of the diff, but is this failing the ingest here? The documentation says that these names will be sanitized if they're in the `bad.files` or `bad.file.extensions` lists. Only the `bad.mime.types` should be rejected

emanns95 replied:
Ah I read a little further. I see ignored files are different from bad.files

emanns95 commented:
I think there needs to be an itest that actually monitors a directory with bad files, bad file extensions, bad file mimetypes, and an ignored file and shows that the first two are sanitized and the last two rejected.

taz77 commented:
build now

mweser commented:
Manual testing verified 100% 👍

johnsliao commented:
build now

johnsliao commented:
Pushed up the following changes:

- Added integration tests
- Moved `Thumbs.db` from `bad.files` to `ignore.files`
- Minor function changes

johnsliao commented:
build now

johnsliao commented:
build now

johnsliao commented:
build now

johnsliao commented:
build now

johnsliao commented:
Summary of pushed changes:
- Fixed up other unit tests
- I-test was flakey when ingesting 2 files via rest endpoint. Broke the itest into two separate itests
- Fixed up documentation

coyotesqrl commented on platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/InputValidation.java:
```diff
@@ -87,7 +95,7 @@ public static String sanitizeFilename(String filename) {
      * @return true if the mime type is acceptable
      */
-    public static boolean checkForClientSideVulnerableMimeType(String mimetype) {
+    public static boolean isMimeTypeClientSideVulnerable(String mimetype) {
```
❗️ This name change makes this method make no sense. The test asks: "is this vulnerable?" and if it is, you return false. If you want to change the name, either make it something like `isMimeTypeClientSideSafe` and keep its current semantics or switch the semantics and update all the calling locations.

mweser replied:
@coyotesqrl I suggested this name change to capture the spirit of using forms of "to be" for boolean-returning methods, but totally agree that the semantics of it were reversed. Good catch. We could always rename it to something like `isMimeTypeClientSideNotVulnerable` to really confuse people :)

coyotesqrl commented on platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/InputValidation.java:
```diff
@@ -97,4 +105,22 @@ public static boolean checkForClientSideVulnerableMimeType(String mimetype) {
+     * @param filename
+     * @return true if the filename is acceptable
+     */
+    public static boolean isFileNameClientSideVulnerable(String filename) {
```
❗️ The same mismatch between method name and method semantics is present in this new method. Either switch the name or switch the semantics.

johnsliao commented:
build now

johnsliao commented:
build now

johnsliao commented:
build now

coyotesqrl commented on platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/InputValidation.java:
```diff
@@ -97,4 +105,22 @@ public static boolean checkForClientSideVulnerableMimeType(String mimetype) {
+     * @return true if the filename is acceptable
+     */
+    public static boolean isFileNameClientSideSafe(String filename) {
+        filename = filename.toLowerCase();
```
✏️ Agree with static analysis recommendation.

johnsliao commented:
That CI build failure is a result of a bad commit and `build now`. I reset hard back to the Job Success CI build.

johnsliao commented:
build now

johnsliao commented:
build now

johnsliao commented:
build now

johnsliao commented:
build now
